src/x/debug/debug.go

```diff
@@ -0,0 +1,108 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package debug
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+)
+
+// Source provides functions for fetching debug information from a single debug source.
+type Source interface {
+	// Write writes it's debug information into the provided writer.
+	Write(w io.Writer) error
+}
+
+// ZipWriter aggregates sources and writes them in a zip file.
+type ZipWriter interface {
+	// WriteZip writes a ZIP file in the provided writer.
+	// The archive contains the dumps of all sources in separate files.
+	WriteZip(io.Writer) error
+	// RegisterSource adds a new source to the produced archive.
+	RegisterSource(string, Source) error
+	// HTTPHandler sends out the ZIP file as raw bytes.
+	HTTPHandler() http.Handler
+	// RegisterHandler wires the HTTPHandlerFunc with the given router.
+	RegisterHandler(string, *http.ServeMux) error
+}
+
+type zipWriter struct {
+	sources map[string]Source
+}
+
+// NewZipWriter returns an instance of an ZipWriter. The passed prefix
+// indicates the folder where to save the zip files.
+func NewZipWriter() ZipWriter {
+	return &zipWriter{
+		sources: make(map[string]Source),
+	}
+}
+
+// RegisterSource adds a new source in the ZipWriter instance.
+// It will return an error if a source with the same filename exists.
+func (i *zipWriter) RegisterSource(dumpFileName string, p Source) error {
+	if _, ok := i.sources[dumpFileName]; ok {
+		return fmt.Errorf("dumpfile already registered %s", dumpFileName)
+	}
+	i.sources[dumpFileName] = p
+	return nil
+}
+
+// WriteZip writes a ZIP file with the data from all sources in the given writer.
+// It will return an error if any of the sources fail to write their data.
+func (i *zipWriter) WriteZip(w io.Writer) error {
+	zw := zip.NewWriter(w)
+	defer zw.Close()
+
+	for filename, p := range i.sources {
+		fw, err := zw.Create(filename)
+		if err != nil {
+			return err
+		}
+		err = p.Write(fw)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (i *zipWriter) HTTPHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		errMsg := "Unable to write ZIP file:"
+		buf := bytes.NewBuffer([]byte{})
+		if err := i.WriteZip(buf); err != nil {
+			http.Error(w, fmt.Sprintf("%s: %s", errMsg, err), http.StatusInternalServerError)
+		}
+		if _, err := io.Copy(w, buf); err != nil {
+			http.Error(w, fmt.Sprintf("%s: %s", errMsg, err), http.StatusInternalServerError)
+		}
```

Comment:
For the error case, I believe it would be good to early return so multiple responses aren't tried to be written out (you can only send HTTP status code header once anyway)?

Also do you want to use our `xhttp.Error(...)` helper?
https://github.com/m3db/m3/blob/master/src/x/net/http/errors.go#L42-L45

And for the second error, perhaps we just log it?  Since the HTTP 200 will have already been sent.

For example:
```go
import (
    // ...
    "github.com/m3db/m3/src/x/instrument"
    "github.com/m3db/m3/src/x/net/http"
    // ...
)

// ...

type zipWriter struct {
  sources map[string]Source
  logger *zap.Logger
}

func NewZipWriter(instrumentOpts instrument.Options) ZipWriter {
  return &zipWriter{
    sources: make(map[string]Source),
    logger: instrumentOpts.Logger(),
  }
}

// ...

func (i *zipWriter) HTTPHandler() http.Handler {
  return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    errMsg := "unable to write ZIP file:"
    buf := bytes.NewBuffer([]byte{})
    if err := i.WriteZip(buf); err != nil {
        xhttp.Error(w, fmt.Sprintf("%s: %s", errMsg, err), http.StatusInternalServerError)
        return
    }
    if _, err := io.Copy(w, buf); err != nil {
        i.logger.Error("unable to write ZIP response", zap.Error(err))
        return
    }
  })
}

// ...
```

Reply:
+1

Reply:
Thanks for the suggestion! Fixed it.